whitesource/component.py

```diff
@@ -84,3 +89,136 @@ def download_component(
+
+    if source_filter:
+        components = _apply_artifact_filter(filter=source_filter, components=components)
+
+    if resource_filter:
+        components = _apply_artifact_filter(filter=resource_filter, components=components)
+
+    return components
+
+
+def _apply_artifact_filter(
+    filter: whitesource.model.WhiteSourceFilterCfg,
+    components: typing.Generator[typing.Union[tuple, typing.Any], typing.Any, typing.Any],
+):
+    logger.debug(f'{filter=}')
+    for c in components:
+        # "source" or "resource", validated earlier
+        artifacts = c.sources if filter.type == 'source' else c.resources
+
+        logger.debug(f'pre filter ({filter.type=}): {len(artifacts)} artifacts')
+
+        if isinstance(m := filter.match, bool):
+            artifacts = list(_apply_artifact_action(
+                action=filter.action,
+                artifacts=artifacts, match=m,
+            ))
+            if filter.type == 'source':
+                c.sources = artifacts
+            elif filter.type == 'resource':
+                c.resources = artifacts
+            logger.debug(f'past filter ({filter.type=}): {len(artifacts)} artifacts')
+            yield c
+
+        elif isinstance(m := filter.match, dict):
+            tmp = []
+            for a in artifacts:
+                if isinstance(a.type, cm.ResourceType):
+                    resource_type_name = a.type.value
+                elif isinstance(a.type, str):
+                    resource_type_name = str(a.type)
+                else:
+                    raise NotImplementedError
+
+                logger.debug(f'parsed {a.type} to {resource_type_name=}')
+
+                if filter.action == 'include':
+                    if re.search(
+                            pattern=m.get('type') if filter.type == 'resource' else m.get('name'),
+                            string=resource_type_name if filter.type == 'resource' else a.name,
+                    ):
+                        tmp.append(a)
+                elif filter.action == 'exclude':
+                    if not re.search(
+                            pattern=m.get('type') if filter.type == 'resource' else m.get('name'),
+                            string=resource_type_name if filter.type == 'resource' else a.name,
+                    ):
+                        tmp.append(a)
+            if filter.type == 'source':
+                c.sources = tmp
+            elif filter.type == 'resource':
+                c.resources = tmp
+            logger.debug(f'past filter ({filter.type=}): {len(artifacts)} artifacts')
+            yield c
+        else:
+            raise NotImplementedError
+
+
+def _apply_artifact_action(
+    action: str,
+    artifacts,
+    match: bool,
+):
+    if action == 'include':
+        return artifacts if match else ()
+    elif action == 'exclude':
+        return artifacts if not match else ()
+    else:
+        raise RuntimeError
```

Comment:
See below (above?) about using RuntimeError.

Also, these two functions behave pretty much the same. Consider somthing like e.g.:
```python3
def _apply_filter_action(
    action: str,
    obj: typing.Union[Foo, Bar], # !
    match: bool,
):
    if action == 'include':
        return obj if match else ()
    elif action == 'exclude':
        return obj if not match else ()
    else:
        raise RuntimeError
```